File: src/IotBbq.App/IotBbq.App/Services/Implementation/ThermometerService.cs

namespace IotBbq.App.Services.Implementation
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;
    using System.Threading.Tasks;
    using Windows.Devices.Gpio;
    using IotBbq.App.Mcp3008;
    using Windows.Devices.Spi;
    using System.Diagnostics;
    using Microsoft.EntityFrameworkCore.Internal;
    using System.Threading;

    public class ThermometerService : IThermometerService
    {
        private const double InputVoltage = 3.3;

        private const double BalancingResistorOhms = 100000;

        private Task initTask;

        private AsyncLock thermometerLock = new AsyncLock();

        private struct CoefficientSet
        {
            public double A;

            public double B;

            public double C;
        }

        /// <summary>
        /// Coefficients
        /// </summary>
        private static readonly CoefficientSet[] Coefficients =
        {
            // These coeificients calculated based on test results
            // 3-19-2018
            // https://www.thinksrs.com/downloads/programs/therm%20calc/ntccalibrator/ntccalculator.html
            // R1 293466 / T1 1.66C
            // R2  96358 / T2 23C
            // R3  42082 / T3 44.44C
            //new CoefficientSet
            //{
            //    A = -1.373357407E-3,
            //    B = 4.914938378E-4,
            //    C = -5.890760444E-7,
            //}

            // When changing these values,
            // don't forget the exponential notation!
            new CoefficientSet
            {
                A = 0.5037245563E-3,
                B = 2.472223870E-4,
                C = 0.07524447351E-7
            }
        };

        private Mcp3008 mcp;

        private static readonly Channel[] Channels =
        {
            Mcp3008.Channels.Single0,
            Mcp3008.Channels.Single1,
            Mcp3008.Channels.Single2,
            Mcp3008.Channels.Single3,
            Mcp3008.Channels.Single4,
            Mcp3008.Channels.Single5,
            Mcp3008.Channels.Single6,
            Mcp3008.Channels.Single7
        };

        public ThermometerService()
        {
            this.mcp = new Mcp3008(0);
            this.initTask = this.mcp.InitializeAsync();
        }

        public async Task<Temps> ReadThermometer(int index)
        {
            if (index < 0 || index > 7)
            {
                throw new IndexOutOfRangeException("Index must be from 0-7");
            }

            using (await this.thermometerLock.LockAsync())
            {
                await this.initTask;

                double sum = 0;
                const int numSamples = 3;
                for (int i = 0; i < numSamples; i++)
                {
                    var reading = this.mcp.Read(Channels[index]);
                    sum += reading.NormalizedValue;
                    await Task.Delay(100);
                }

                double averageValue = sum / numSamples;

                //Debug.WriteLine($"Reading for thermometer {index} is {reading.RawValue}, Normalized: {reading.NormalizedValue}");

                double voltage = averageValue * InputVoltage;
                double resistance = TempUtils.GetThermistorResistenceFromVoltage(3.3, voltage, BalancingResistorOhms);
                //Debug.WriteLine($"Got Resistance {resistance} from voltage {voltage}");

                CoefficientSet coefficients = Coefficients[0];

                var temps = new Temps();
                if (double.IsInfinity(resistance))
                {
                    temps.Kelvin = 0;
                }
                else
                {
                    temps.Kelvin = TempUtils.ResistanceToTemp(coefficients.A, coefficients.B, coefficients.C, resistance);
                }

                temps.Celcius = TempUtils.KelvinToCelcius(temps.Kelvin);
                temps.Farenheight = TempUtils.CelciusToFarenheight(temps.Celcius);

                return temps;
            }
        }
    }
}

File: src/IotBbq.App/IotBbq.JsApp/src/app/services/ThermometerService.ts
import { Inject } from '@angular/core';
import { SPICLIENT_TOKEN, ISpiClient } from './contracts/ISpiClient';
import { Mcp3008, Channels, Channel } from './Mcp3008';
import { TempUtils } from './TempUtils';
import { Utility } from './Utility';

export enum ThermometerState {
  Disconnected,
  Connected,
}

export interface ITemps {
  state: ThermometerState;
  kelvin: number;
  celcius: number;
  farenheight: number;
}

interface CoefficientSet {
  A: number;
  B: number;
  C: number;
}

export class ThermometerService {

  private static inputVoltage = 3.3;

  private static balancingResistorOhms = 100000;

  private mcp: Mcp3008;

  private channels: Channel[] = [
    Channels.Single0,
    Channels.Single1,
    Channels.Single2,
    Channels.Single3,
    Channels.Single4,
    Channels.Single5,
    Channels.Single6,
    Channels.Single7,
  ];

  private coefficients: CoefficientSet = {
    A: 0.5037245563E-3,
    B: 2.472223870E-4,
    C: 0.07524447351E-7,
  };

  constructor(
    @Inject(SPICLIENT_TOKEN) spiClient: ISpiClient
  ) {
    this.mcp = new Mcp3008(spiClient, 0);
  }

  public async readThermometer(probeNumber: number): Promise<ITemps> {

    if (probeNumber < 1 || probeNumber > 8) {
      throw new Error('Probe number must be from 1-8');
    }

    // SPI is zero-based so adjust probe number to the appropriate index
    const probeIndex = probeNumber - 1;

    let sum = 0;
    const numSamples = 3;
    for (let i = 0; i < numSamples; i++) {
      const reading = await this.mcp.read(this.channels[probeIndex]);

      // If any of our reading are zero, we have to treat it as disconnected
      if (reading.getRawValue() === 0) {
        return {
          state: ThermometerState.Disconnected,
          celcius: 0,
          farenheight: 0,
          kelvin: 0,
        };
      }

      sum += reading.getNormalizedValue();
      await Utility.sleep(100);
    }

    // We were able to read all X samples, compute the average
    const averageValue = sum / numSamples;

    const voltage = averageValue * ThermometerService.inputVoltage;
    const resistance = TempUtils.getThermistorResistenceFromVoltage(3.3, voltage, ThermometerService.balancingResistorOhms);

    const temps: ITemps = {
      state: ThermometerState.Disconnected,
      celcius: 0,
      farenheight: 0,
      kelvin: 0,
    };

    // TODO: Check this check, not sure how this would be possible
    if (resistance !== NaN) {
      temps.state = ThermometerState.Connected;
      temps.kelvin = TempUtils.resistanceToTemp(this.coefficients.A, this.coefficients.B, this.coefficients.C, resistance);  
      temps.celcius = TempUtils.kelvinToCelcius(temps.kelvin);
      temps.farenheight = TempUtils.celciusToFarenheight(temps.celcius);
    }

    return temps;
  }
}
